pymemcache/client/hash.py

```diff
@@ -0,0 +1,282 @@
+            self.add_server(server, port)
+
+    def add_server(self, server, port):
+        key = '%s:%s' % (server, port)
+
+        if self.use_pooling:
+            client = PooledClient(
+                (server, port),
+                **self.default_kwargs
+            )
+        else:
+            client = Client((server, port))
+
+        self.clients[key] = client
+        self.hasher.add_node(key)
+
+    def remove_server(self, server, port):
+        dead_time = time.time()
+        self._failed_clients.pop((server, port))
+        self._dead_clients[(server, port)] = dead_time
+        key = '%s:%s' % (server, port)
+        self.hasher.remove_node(key)
+
+    def _get_client(self, key):
+        _check_key(key, self.key_prefix)
+        if len(self._dead_clients) > 0:
+            current_time = time.time()
+            ldc = self._last_dead_check_time
+            # we have dead clients and we have reached the
+            # timeout retry
+            if current_time - ldc > self.dead_timeout:
+                for server, dead_time in self._dead_clients.items():
+                    if current_time - dead_time > self.dead_timeout:
+                        print('bringing server back in rotation')
+                        self.add_server(*server)
+                        self._last_dead_check_time = current_time
+
+        server = self.hasher.get_node(key)
+        print('got server %s' % server)
+        client = self.clients[server]
+        return client
+
+    def _safely_run_func(self, client, func, *args, **kwargs):
+        try:
+            if client.server in self._failed_clients:
+                # This server is currently failing, lets check if it is in
+                # retry or marked as dead
+                failed_metadata = self._failed_clients[client.server]
+
+                # we haven't tried our max amount yet, if it has been enough
+                # time lets just retry using it
+                if failed_metadata['attempts'] < self.retry_attempts:
+                    failed_time = failed_metadata['failed_time']
+                    if time.time() - failed_time > self.retry_timeout:
+                        print(failed_metadata)
+                        print('retrying')
+                        result = func(*args, **kwargs)
+                        # we were successful, lets remove it from the failed
+                        # clients
+                        self._failed_clients.pop(client.server)
+                        return result
+                    return
+                else:
+                    # We've reached our max retry attempts, we need to mark
+                    # the sever as dead
+                    print('marking as dead')
+                    self.remove_server(*client.server)
+
+            result = func(*args, **kwargs)
+            return result
+
+        # Connecting to the server fail, we should enter
+        # retry mode
+        except socket.error:
```

Comment:
I think I must be misunderstanding something about your comment or the code, so let me over-explain what I was trying to say so we can get on the same page.

Consider this code, using `HashClient`:

```
client = HashClient([("server1", 11211), ("server2", 11211), ("server3", 11211)])

val = client.get("my_key")
if val is None:
    val = my_service.some_method(some_args)
    client.set("my_key", val)
do_something(val)
```

This is a pretty typical interaction with a memcache client. Let's assume that HashClient has exactly the same contract for `get` and `set` (with `ignore_exc=False` by default). In that case, if "server1", "server2" and "server3" are all down, we would expect `client.get()` to raise an Exception. If the servers were available when `get` was called, but went down before `set` was called, then we would expect `set` to raise an Exception. If `ignore_exc=True`, then we would expect `get` to never raise (it would just return `None` when the servers were down), but `set` would still raise if the servers were unavailable.

What I'm pointing out is that the actual contract for `HashClient` has it always returning `None` from both `get` and `set` if the hashed server is down (even if some of the other servers are up, as we don't do retries on socket errors). That means `HashClient` acts like `ignore_exc=True` for both `get` and `set` methods at all times (and is not configurable). It's actually a little bit worse than that, as `_run_cmd()` always returns `None` when there is a socket error, and that breaks the contract of functions like `cas` and `get_many`, which either have a specific meaning for `None`, or should not ever return it (respectively). The implementation of `ignore_exc` in `Client` is careful to do the right thing for those cases.

There's no reason to choose between gracefully handling failures and supporting the `Client` contract. It is easy to do both. If you are uncomfortable with `ignore_exc`, then the simple way to do both is to add these lines:

```
...
except socket.error:
    exc_info = sys.exc_inf()
    ... your current failure logic ...
    raise exc_info[0], exc_info[1], exc_info[2]
```

If you would like to support the `ignore_exc` feature in `HashingClient` you'll have to wrap the last line of that in a conditional that checks the memcache call name and the value of ignore_exc and decides whether to return a value (we return `{}` in Client in that case) or raise.

FWIW, the `ignore_exc` feature is widely used (from what other users have told me), and is turned on by default at Pinterest, mostly because we do not want memcache server outages to cause site outages, and we don't want to wrap all calls to `get` in try/except clauses. I think it is worth supporting for `HashClient`, but I don't consider it a blocking change.